truckersmp-cli

```diff
@@ -1,512 +1,547 @@
-                -O "$moddir/$f" "$dlurl/$f" \
-            || wget --header 'Accept-encoding: identity' -v \
-                -O "$moddir/$f" "$dlurlalt/$f" \
-            || error=1
-        fi
-    done
-
-    # something went wrong
-    if [ "$error" = "1" ]; then
-        >&2 echo "Failed to download mod files."
-        quit 1
-    fi
-}
-
-update_game() {
+    modfiles = []
+    try:
+        for item in json.JSONDecoder().decode(files_json.decode("ascii"))["Files"]:
+            modfiles.append((item["Md5"], item["FilePath"]))
+        if len(modfiles) == 0:
+            raise Exception("File list is empty")
+    except Exception as e:
+        sys.exit("""Failed to parse files.json: {}
+Please report an issue: {}""".format(e, URL.issueurl))
+
+    # compare existing local files with md5sums and remember missing/wrong files
+    files_to_download = []
+    for md5, jsonfilepath in modfiles:
+        md5hash = hashlib.md5()
+        modfilepath = os.path.join(args.moddir, jsonfilepath[1:])
+        if not os.path.isfile(modfilepath):
+            files_to_download.append(("/files" + jsonfilepath, modfilepath, md5))
+        else:
+            try:
+                with open(modfilepath, "rb") as f:
+                    while True:
+                        buf = f.read(md5hash.block_size * 4096)
+                        if not buf:
+                            break
+                        md5hash.update(buf)
+                if md5hash.hexdigest() != md5:
+                    files_to_download.append(("/files" + jsonfilepath, modfilepath, md5))
+            except Exception as e:
+                sys.exit("Failed to read {}: {}".format(modfilepath, e))
+    if len(files_to_download) > 0:
+        printdbg("Files to download:")
+        for path, _, _ in files_to_download:
+            printdbg("  {}".format(path))
+    else:
+        printdbg("No files to download")
+
+    # download missing/wrong files
+    if not download_files(URL.dlurl, files_to_download):
+        if not download_files(URL.dlurlalt, files_to_download):
+            # something went wrong
+            sys.exit("Failed to download mod files.")
+
+def update_game():
     # make sure steam is closed before updating
     # it's possible to update with the steam client open but the client looses
     # all connectivity and asks for password and steam guard code after restart
-    while pidof steam > /dev/null 2>&1; do
-        print "found running steam, running pkill -x steam"
-        pkill -x steam
-    sleep 5
-    done
-
-    if ! [ -d "$gamedir" ]; then
-        print "mkdir $gamedir"
-        mkdir -p "$gamedir"
-    fi
+    while True:
+        try:
+            subprocess.check_call(["pidof", "steam"], stdout = subprocess.DEVNULL)
+        except:
+            break
+        else:
+            printdbg("Found running steam, running pkill -x steam")
+            subprocess.call(["pkill", "-x", "steam"])
+            time.sleep(5)
```

Comment:
Maybe we could do better than just waiting. I think we can use `os.kill` directly in a loop to test for process termination (os.waitpid only work with child processes). Also we could get rid of the subprocess call.

cf. https://stackoverflow.com/a/7654102